File: sdk/js/src/client.test.ts
import { describe, it, expect, beforeAll, afterAll } from "vitest";
import { createServer, type Server } from "node:http";
import { AutoclawClient } from "./client.js";

const fakeStatus = {
  running: true,
  total_experiments: 3,
  best_score: 0.87,
  budget_remaining: 142.5,
  uptime: 200.0,
};

const fakeExperiments = [
  { id: "exp-001", hypothesis: "lr=2e-5", params: {}, metrics: {},
    score: 0.82, status: "completed", timestamp: "2026-06-18T10:00:00Z",
    git_hash: "abc", duration_seconds: 12 },
  { id: "exp-002", hypothesis: "lr=3e-5", params: {}, metrics: {},
    score: 0.87, status: "completed", timestamp: "2026-06-18T10:01:00Z",
    git_hash: "def", duration_seconds: 14 },
];

let server: Server;
let url: string;

beforeAll(async () => {
  server = createServer((req, res) => {
    res.setHeader("Content-Type", "application/json");
    if (req.method === "GET" && req.url === "/api/status")
      return res.end(JSON.stringify(fakeStatus));
    if (req.method === "GET" && req.url === "/api/results")
      return res.end(JSON.stringify(fakeExperiments));
    if (req.method === "GET" && req.url === "/api/context") {
      res.setHeader("Content-Type", "text/plain");
      return res.end("# MISSION\nTest.\n");
    }
    if (req.method === "POST" && ["/api/start","/api/stop","/api/reset"].includes(req.url!))
      return res.end(JSON.stringify({ status: req.url!.split("/").pop() }));
    if (req.method === "POST" && req.url === "/api/context")
      return res.end(JSON.stringify({ status: "updated" }));
    res.statusCode = 404; res.end();
  });
  await new Promise<void>(r => server.listen(0, "127.0.0.1", () => r()));
  const addr = server.address();
  url = typeof addr === "object" && addr ? `http://127.0.0.1:${addr.port}` : "";
});

afterAll(() => new Promise<void>(r => server.close(() => r())));

describe("AutoclawClient", () => {
  it("status round-trip", async () => {
    const c = new AutoclawClient({ baseUrl: url });
    const s = await c.status();
    expect(s.best_score).toBe(0.87);
    expect(s.running).toBe(true);
  });

  it("experiments typed", async () => {
    const c = new AutoclawClient({ baseUrl: url });
    const exps = await c.experiments();
    expect(exps).toHaveLength(2);
    expect(exps[1].score).toBe(0.87);
  });

  it("best picks highest", async () => {
    const c = new AutoclawClient({ baseUrl: url });
    const best = await c.best();
    expect(best?.id).toBe("exp-002");
  });

  it("context round-trip", async () => {
    const c = new AutoclawClient({ baseUrl: url });
    expect(await c.getContext()).toContain("MISSION");
    await c.setContext("# new\n");
  });

  it("lifecycle endpoints", async () => {
    const c = new AutoclawClient({ baseUrl: url });
    expect((await c.start()).status).toBe("start");
    expect((await c.stop()).status).toBe("stop");
    expect((await c.reset()).status).toBe("reset");
  });
});

File: sdk/go/client_test.go
package autoclaw

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

func fakeServer(t *testing.T) *httptest.Server {
	t.Helper()
	mux := http.NewServeMux()
	mux.HandleFunc("/api/status", func(w http.ResponseWriter, _ *http.Request) {
		json.NewEncoder(w).Encode(Status{Running: true, TotalExperiments: 3, BestScore: 0.87, BudgetRemaining: 142.5, Uptime: 200})
	})
	mux.HandleFunc("/api/results", func(w http.ResponseWriter, _ *http.Request) {
		json.NewEncoder(w).Encode([]Experiment{
			{ID: "exp-001", Hypothesis: "lr=2e-5", Score: 0.82, Status: "completed"},
			{ID: "exp-002", Hypothesis: "lr=3e-5", Score: 0.87, Status: "completed"},
		})
	})
	mux.HandleFunc("/api/context", func(w http.ResponseWriter, r *http.Request) {
		if r.Method == "GET" {
			w.Header().Set("Content-Type", "text/plain")
			w.Write([]byte("# MISSION\nTest.\n"))
			return
		}
		json.NewEncoder(w).Encode(map[string]string{"status": "updated"})
	})
	for _, p := range []string{"/api/start", "/api/stop", "/api/reset"} {
		path := p
		mux.HandleFunc(path, func(w http.ResponseWriter, _ *http.Request) {
			json.NewEncoder(w).Encode(map[string]string{"status": strings.TrimPrefix(path, "/api/")})
		})
	}
	return httptest.NewServer(mux)
}

func TestStatus(t *testing.T) {
	srv := fakeServer(t)
	defer srv.Close()
	c := NewClient(srv.URL)
	s, err := c.Status(context.Background())
	if err != nil {
		t.Fatal(err)
	}
	if s.BestScore != 0.87 {
		t.Errorf("BestScore = %v, want 0.87", s.BestScore)
	}
	if !s.Running {
		t.Error("expected Running=true")
	}
}

func TestExperiments(t *testing.T) {
	srv := fakeServer(t)
	defer srv.Close()
	c := NewClient(srv.URL)
	es, err := c.Experiments(context.Background())
	if err != nil {
		t.Fatal(err)
	}
	if len(es) != 2 {
		t.Fatalf("len = %d, want 2", len(es))
	}
	if es[1].Score != 0.87 {
		t.Errorf("es[1].Score = %v, want 0.87", es[1].Score)
	}
}

func TestBest(t *testing.T) {
	srv := fakeServer(t)
	defer srv.Close()
	c := NewClient(srv.URL)
	best, err := c.Best(context.Background())
	if err != nil || best == nil {
		t.Fatalf("Best returned err=%v, best=%v", err, best)
	}
	if best.ID != "exp-002" {
		t.Errorf("best.ID = %q, want exp-002", best.ID)
	}
}

func TestContextRoundTrip(t *testing.T) {
	srv := fakeServer(t)
	defer srv.Close()
	c := NewClient(srv.URL)
	text, err := c.GetContext(context.Background())
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(text, "MISSION") {
		t.Errorf("context missing MISSION: %q", text)
	}
	if err := c.SetContext(context.Background(), "# new\n"); err != nil {
		t.Errorf("SetContext: %v", err)
	}
}

func TestLifecycle(t *testing.T) {
	srv := fakeServer(t)
	defer srv.Close()
	c := NewClient(srv.URL)
	ctx := context.Background()
	for _, fn := range []func(context.Context) error{c.Start, c.Stop, c.Reset} {
		if err := fn(ctx); err != nil {
			t.Errorf("lifecycle: %v", err)
		}
	}
}
